web_poet/rules.py

```diff
@@ -0,0 +1,338 @@
+from __future__ import annotations  # https://www.python.org/dev/peps/pep-0563/
+
+import importlib
+import importlib.util
+import pkgutil
+import warnings
+from collections import deque
+from operator import attrgetter
+from typing import Any, Dict, Iterable, List, Optional, Type, TypeVar, Union
+
+import attrs
+from url_matcher import Patterns
+
+from web_poet._typing import get_item_cls
+from web_poet.pages import ItemPage
+from web_poet.utils import _create_deprecated_class, as_list, str_to_pattern
+
+Strings = Union[str, Iterable[str]]
+
+PageObjectRegistryTV = TypeVar("PageObjectRegistryTV", bound="PageObjectRegistry")
+
+
+@attrs.define(frozen=True)
+class ApplyRule:
+    """A rule that primarily applies Page Object and Item overrides for a given
+    URL pattern.
+
+    This is instantiated when using the :func:`web_poet.handle_urls` decorator.
+    It's also being returned as a ``List[ApplyRule]`` when calling the
+    ``web_poet.default_registry``'s :meth:`~.PageObjectRegistry.get_rules`
+    method.
+
+    You can access any of its attributes:
+
+        * ``for_patterns`` - contains the list of URL patterns associated with
+          this rule. You can read the API documentation of the `url-matcher
+          <https://url-matcher.readthedocs.io/>`_ package for more information
+          about the patterns.
+        * ``use`` - The Page Object that will be **used** in cases where the URL
+          pattern represented by the ``for_patterns`` attribute is matched.
+        * ``instead_of`` - *(optional)* The Page Object that will be **replaced**
+          with the Page Object specified via the ``use`` parameter.
+        * ``to_return`` - *(optional)* The data container class that will be
+          **replaced** by the item returned by the Page Object specified via the
+          ``use`` parameter.
```

Comment:
I think we should try to find a better wording. Unlike Page Objects, we're not replacing anything here. The idea is to mark a Page Object as something which can return an item of a certain type. Moreover, `to_return` is for an advanced use case where this declaration needs to be different from the item type which is really returned by the Page Object; most users should rely on automatic behavior.